File: android/app/src/main/java/com/zulipmobile/notifications/FcmListenerService.java
package com.zulipmobile.notifications;

import android.content.Context;
import com.facebook.react.ReactApplication;
import com.facebook.react.bridge.ReactContext;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import com.zulipmobile.MainApplication;

public class FcmListenerService extends FirebaseMessagingService {
    @Override
    public void onMessageReceived(RemoteMessage message) {
        final Context applicationContext = getApplicationContext();
        if (!(applicationContext instanceof MainApplication)) {
            return;
        }
        NotificationUiManager.onReceived(this, message.getData());
    }

    @Override
    public void onNewToken(String token) {
        super.onNewToken(token);
        final ReactContext reactContext =
                ((ReactApplication) getApplication())
                        .getReactNativeHost()
                        .getReactInstanceManager()
                        .getCurrentReactContext();
        NotificationsModule.emitToken(reactContext, token);
    }
}

File: android/app/src/main/java/com/zulipmobile/notifications/FcmListenerService.kt
package com.zulipmobile.notifications

import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.facebook.react.bridge.ReactContext
import com.facebook.react.ReactApplication

class FcmListenerService : FirebaseMessagingService() {
    override fun onMessageReceived(message: RemoteMessage) {
        onReceived(this, message.data)
    }

    override fun onNewToken(token: String) {
        super.onNewToken(token)
        val reactContext = (application as ReactApplication)
            .reactNativeHost
            .reactInstanceManager
            .currentReactContext
        NotificationsModule.emitToken(reactContext, token)
    }
}
